File: sdks/python/durable_sdk/tool_server.py
"""Tool server that runs as a child process of the Rust runtime.

Reads tool execution requests from stdin, dispatches to registered
handlers, writes results to stdout. Speaks the durable NDJSON protocol.
"""

import json
import sys
import traceback
from typing import Any, Callable, Dict, Optional

from .protocol import ProtocolMessage, read_message, write_message


class DurableToolServer:
    """Register tool handlers and run the stdin/stdout dispatch loop.

    Example:
        server = DurableToolServer()

        @server.tool("greet", "Greet someone by name")
        def greet(name: str = "world"):
            return f"Hello, {name}!"

        server.run()
    """

    def __init__(self):
        self._handlers: Dict[str, Callable] = {}
        self._definitions: list = []

    def tool(
        self,
        name: str,
        description: str,
        parameters: Optional[dict] = None,
    ):
        """Decorator to register a tool handler."""
        def decorator(fn: Callable):
            self._handlers[name] = fn
            self._definitions.append({
                "name": name,
                "description": description,
                "parameters": parameters or {
                    "type": "object",
                    "properties": {},
                },
            })
            return fn
        return decorator

    def register(
        self,
        name: str,
        handler: Callable,
        description: str = "",
        parameters: Optional[dict] = None,
    ):
        """Register a tool handler imperatively."""
        self._handlers[name] = handler
        self._definitions.append({
            "name": name,
            "description": description,
            "parameters": parameters or {"type": "object", "properties": {}},
        })

    def run(self):
        """Run the tool server loop. Reads from stdin, writes to stdout."""
        while True:
            msg = read_message()
            if msg is None:
                break  # EOF

            if msg.type == "execute":
                tool_name = msg.payload.get("tool_name", "")
                arguments = msg.payload.get("arguments", {})
                self._handle_execute(tool_name, arguments)
            elif msg.type == "heartbeat":
                ts = msg.payload.get("timestamp", 0)
                write_message(ProtocolMessage("heartbeat_ack", timestamp=ts))
            else:
                write_message(ProtocolMessage(
                    "error",
                    message=f"unknown message type: {msg.type}",
                    retryable=False,
                ))

    def _handle_execute(self, tool_name: str, arguments: Any):
        handler = self._handlers.get(tool_name)
        if handler is None:
            write_message(ProtocolMessage(
                "error",
                message=f"unknown tool: {tool_name}",
                retryable=False,
            ))
            return

        try:
            if isinstance(arguments, dict):
                result = handler(**arguments)
            else:
                result = handler(arguments)
            write_message(ProtocolMessage("result", output=result))
        except Exception as e:
            write_message(ProtocolMessage(
                "error",
                message=str(e),
                retryable=False,
            ))

File: sdks/typescript/src/tool-server.ts
/**
 * Tool server that speaks the durable NDJSON protocol on stdio.
 * Zero dependencies — Node.js built-ins only.
 */

import { createMessageReader, writeMessage, ProtocolMessage } from "./protocol";

type ToolHandler = (args: Record<string, unknown>) => unknown | Promise<unknown>;

interface RegisteredTool {
  name: string;
  description: string;
  handler: ToolHandler;
  parameters?: Record<string, unknown>;
}

/**
 * Register tool handlers and run the stdin/stdout dispatch loop.
 *
 * @example
 * ```ts
 * const server = new DurableToolServer();
 * server.register("greet", "Greet someone", (args) => `Hello, ${args.name}!`);
 * server.run();
 * ```
 */
export class DurableToolServer {
  private tools = new Map<string, RegisteredTool>();

  register(
    name: string,
    description: string,
    handler: ToolHandler,
    parameters?: Record<string, unknown>
  ): void {
    this.tools.set(name, { name, description, handler, parameters });
  }

  async run(): Promise<void> {
    for await (const msg of createMessageReader()) {
      if (msg.type === "execute") {
        const toolName = msg.tool_name as string;
        const args = (msg.arguments as Record<string, unknown>) || {};
        await this.handleExecute(toolName, args);
      } else if (msg.type === "heartbeat") {
        writeMessage({ type: "heartbeat_ack", timestamp: msg.timestamp });
      } else {
        writeMessage({
          type: "error",
          message: `unknown message type: ${msg.type}`,
          retryable: false,
        });
      }
    }
  }

  private async handleExecute(
    toolName: string,
    args: Record<string, unknown>
  ): Promise<void> {
    const tool = this.tools.get(toolName);
    if (!tool) {
      writeMessage({
        type: "error",
        message: `unknown tool: ${toolName}`,
        retryable: false,
      });
      return;
    }

    try {
      const result = await tool.handler(args);
      writeMessage({ type: "result", output: result });
    } catch (err: unknown) {
      writeMessage({
        type: "error",
        message: err instanceof Error ? err.message : String(err),
        retryable: false,
      });
    }
  }
}
